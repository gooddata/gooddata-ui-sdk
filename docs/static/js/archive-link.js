document.addEventListener('DOMContentLoaded', function () {
    function createArchiveLink() {
        const link = document.createElement('a');
        link.className = 'dropdown-item';
        
        const currentHostname = window.location.hostname;
        if (currentHostname.includes('netlify.app')) {
            // For Netlify previews, link to the versions page within the current preview site
            link.href = window.location.origin + '/docs/gooddata-ui/versions/';
        } else if (currentHostname === 'localhost' || currentHostname === '127.0.0.1') {
            // For local testing, link to the versions page
            link.href = '/versions/';
        } else {
            // For production - link to the unversioned archive page
            link.href = 'https://www.gooddata.com/docs/gooddata-ui/versions/';
        }
        
        // Safe DOM manipulation - XSS protected
        const icon = document.createElement('i');
        icon.className = 'fas fa-archive me-2';
        icon.setAttribute('aria-hidden', 'true');
        
        link.appendChild(icon);
        link.appendChild(document.createTextNode('Archive'));
        
        // Add descriptive text for screen readers
        link.setAttribute('aria-label', 'View all versions in archive');
        
        return link;
    }
    
    // Select the dropdown menu and add the archive link
    const dropdownMenus = document.getElementsByClassName('dropdown-menu');
    for (let i = 0; i < dropdownMenus.length; i++) {
        const archiveLink = createArchiveLink();
        dropdownMenus[i].appendChild(archiveLink);
    }
});