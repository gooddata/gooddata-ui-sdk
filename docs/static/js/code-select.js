(function ($) {
    $(document).ready(function () {
        $(".gd-docs-code-select__tab").on("click", setActive);
        $(".gd-docs-code-select__osdetection .gd-docs-code-select__tab").each(osDetection);
    });

    var setActive = function() {
        if(!$(this).hasClass("active")) {
            $(this).siblings().removeClass("active");
            $(this)
                .addClass("active")
                .parent()
                .parent()
                .find(".gd-docs-code-select__code")
                .removeClass("active")
                .filter(".lang-" + $(this).data("lang"))
                .addClass("active");
        }
    }

    var osDetection = function() {
        if(window.navigator.userAgent.indexOf($(this).data("user-agent")) >= 0) {
            $(this).trigger("click");
        }
    }

}(jQuery));