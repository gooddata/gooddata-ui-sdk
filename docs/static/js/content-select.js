(function ($) {
    $(document).ready(function () {
        $(".gd-docs-content-select__tab").on("click", setActive);
    });

    var setActive = function() {
        if(!$(this).hasClass("active")) {
            $(this).siblings().removeClass("active");

            $(this).addClass("active")
                .parent()
                .parent()
                .find(".gd-docs-content__block")
                .removeClass("active")
                .filter(".content-" + $(this).data("content"))
                .addClass("active");
        }
    }
}(jQuery));
