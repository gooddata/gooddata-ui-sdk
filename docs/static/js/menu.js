$(document).ready(function () {
    var menu = $(".gd-docs-menu"),
        menuActive = $(".gd-docs-menu-page.active");

    if(menuActive.length > 0) {
        menu.scrollTop(menuActive.offset().top - menu.offset().top - parseInt(menu.css("paddingTop")));
    }

    menu.removeClass("init");
});