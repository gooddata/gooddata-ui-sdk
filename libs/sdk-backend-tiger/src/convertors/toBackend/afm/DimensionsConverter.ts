// (C) 2007-2024 GoodData Corporation
import {
    Dimension,
    SortKeyAttribute,
    SortKeyAttributeAttributeDirectionEnum,
    SortKeyAttributeAttributeSortTypeEnum,
    SortKeyValue,
} from "@gooddata/api-client-tiger";
import {
    IExecutionDefinition,
    ILocatorItem,
    isAttributeAreaSort,
    isAttributeLocator,
    isMeasureLocator,
    isAttributeSort,
    ISortItem,
    MeasureGroupIdentifier,
    SortDirection,
} from "@gooddata/sdk-model";
import isEmpty from "lodash/isEmpty.js";
import findIndex from "lodash/findIndex.js";

type SortKey = SortKeyAttribute | SortKeyValue;

function merge(dims: Dimension[], sorting: SortKey[][]): Dimension[] {
    return dims.map((dim, idx) => {
        if (!isEmpty(sorting[idx])) {
            return {
                ...dim,
                sorting: sorting[idx],
            };
        }

        return dim;
    });
}

function convertSortDirection(direction: SortDirection): SortKeyAttributeAttributeDirectionEnum {
    if (direction === "asc") {
        return SortKeyAttributeAttributeDirectionEnum.ASC;
    }

    return SortKeyAttributeAttributeDirectionEnum.DESC;
}

/**
 * For normal attribute sort by value:
 * - DEFAULT sort type is always used.
 * It means, that potential default sort label on model will have always higher priority over sort label provided by execution.
 * It behaves similar to the Bear behavior and according UX but it limits API capabilities as API can override default sort label by the one from execution by using SortType.LABEL
 * If someone needs this here too, we need to extend IAttributeSortItem by some label prioritization flag and consider it in this convertor.
 *
 * For area sort it just uses SortType.AREA to use this sort by numerical sum in execution.
 */
function convertAttributeSortType(sortItem: ISortItem): SortKeyAttributeAttributeSortTypeEnum {
    return isAttributeAreaSort(sortItem)
        ? SortKeyAttributeAttributeSortTypeEnum.AREA
        : SortKeyAttributeAttributeSortTypeEnum.DEFAULT;
}

/**
 * This messed up function exists because PivotTable expects specific (bear) format of attribute element URIs from
 * which it construct table column ids. The logic behind column ids is deeply rooted (rotted-into) that it would
 * mean non-trivial changes / partial rewrites in the pivot table.
 *
 * To avoid that (because it will likely be several weeks worth of effort and testing), we opted for a dirty 'trick'
 * where tiger backend constructs attribute element URIs so that they resemble the bear URIs. The element ID is
 * actually the primaryLabelValue -â†’ thing that can be used for sorting.
 *
 * This function caters for the dirty trick + in case the sorts were created 'normally', programmatically by the
 * user who specified primaryLabelValue directly, it has logic to fall back to using the uri as-is.
 */
function extractItemValueFromElement(elementUri: string | null): string | null {
    // no reasonable way to avoid the super-linear backtracking right now
    // eslint-disable-next-line regexp/no-super-linear-backtracking, regexp/no-unused-capturing-group
    const parsedUri = elementUri?.match(/obj\/([^/]*)(\/elements\?id=)?(.*)$/);

    if (parsedUri?.[3]) {
        return parsedUri[3];
    }

    return elementUri;
}

function convertMeasureLocators(locators: ILocatorItem[]): { [key: string]: string | null } {
    const dataColumnLocators = locators.map<{ [key: string]: string | null }>((locator) => {
        if (isAttributeLocator(locator)) {
            return {
                [locator.attributeLocatorItem.attributeIdentifier]: extractItemValueFromElement(
                    locator.attributeLocatorItem.element,
                ),
            };
        } else if (isMeasureLocator(locator)) {
            return {
                [MeasureGroupIdentifier]: locator.measureLocatorItem.measureIdentifier,
            };
        } else {
            return {};
        }
    });
    return Object.assign({}, ...dataColumnLocators);
}

/**
 * Places sorting into dimensions. Returns new dimensions augmented by sorting. Does not mutate.
 *
 * Tiger does sorting differently from bear so this is somewhat more complicated than pure object conversions.
 *
 * 1. Sorting is now placed in the dimension that has to be sorted
 * 2. When sorting by attribute (headers), then the attribute sort key must be placed into the dimension that
 *    contains the attribute
 * 3. When sorting by measure, we now fall back to 'bear-like' behavior: the dimension opposite to the one
 *    that contains the measures will be sorted. It will be sorted using the measure (possibly scoped for
 *    particular attribute values) located in the MeasureGroup dimension.
 *
 * At the end, this function walks the sort items in the order defined by the user and distributes them into
 * the dimensions. The function is lenient for now and will log warnings and ignore anything weird that it
 * cannot process (e.g. measure sorting when there is no dim with MeasureGroup, or if there is only dim with MeasureGroup
 * and no other dim).
 *
 * @param dims - dimensions to add sorting to
 * @param sorts - sort items defined by SDK user
 */
function dimensionsWithSorts(dims: Dimension[], sorts: ISortItem[]): Dimension[] {
    if (isEmpty(sorts)) {
        return dims;
    }

    const nonMeasureDimIdx = findIndex(dims, (dim) => !dim.itemIdentifiers.includes(MeasureGroupIdentifier));
    const measureDim = dims.find((dim) => dim.itemIdentifiers.includes(MeasureGroupIdentifier));
    const sorting: SortKey[][] = dims.map((_) => []);

    sorts.forEach((sortItem) => {
        if (isAttributeSort(sortItem)) {
            const attributeIdentifier = sortItem.attributeSortItem.attributeIdentifier;
            const direction = convertSortDirection(sortItem.attributeSortItem.direction);

            const attributeSortKey: SortKeyAttribute = {
                attribute: {
                    attributeIdentifier,
                    // ASC direction is default for Tiger backend. Dont send it to execution to prevent override of possible default sort label direction. It has the same meaning as TigerSortDirection.DEFAULT which does not exist.
                    ...(direction !== SortKeyAttributeAttributeDirectionEnum.ASC ? { direction } : {}),
                    sortType: convertAttributeSortType(sortItem),
                },
            };

            const dimIdx = findIndex(dims, (dim) => dim.itemIdentifiers.includes(attributeIdentifier));

            if (dimIdx < 0) {
                // eslint-disable-next-line no-console
                console.log(
                    `attempting to sort by attribute with localId ${attributeIdentifier} but this attribute is not in any dimension.`,
                );

                return;
            }

            sorting[dimIdx].push(attributeSortKey);
        } else {
            if (nonMeasureDimIdx < 0) {
                console.warn(
                    "Trying to use measure sort in an execution that only contains dimension with MeasureGroup. " +
                        "This is not valid sort. Measure sort is used to sort the non-measure dimension by values from measure dimension. Ignoring",
                );

                return;
            }

            if (!measureDim) {
                console.warn(
                    "Trying to use measure sort in an execution that does not contain MeasureGroup. Ignoring.",
                );

                return;
            }

            const valueSortKey: SortKeyValue = {
                value: {
                    direction: convertSortDirection(sortItem.measureSortItem.direction),
                    dataColumnLocators: {
                        [measureDim.localIdentifier!]: convertMeasureLocators(
                            sortItem.measureSortItem.locators,
                        ),
                    },
                },
            };

            sorting[nonMeasureDimIdx].push(valueSortKey);
        }
    });

    return merge(dims, sorting);
}

export function dimensionLocalIdentifier(idx: number): string {
    return `dim_${idx}`;
}

/**
 * Converts data in execution definition into dimension specifications for tiger. The tiger specifics
 * are that dimensions have localIds and that sorting is specified per-dimension.
 *
 * @param def - execution definition to convert
 */
export function convertDimensions(def: IExecutionDefinition): Dimension[] {
    const dimensionsWithoutSorts: Dimension[] = def.dimensions.map((dim, idx) => {
        return {
            localIdentifier: dimensionLocalIdentifier(idx),
            itemIdentifiers: dim.itemIdentifiers,
        };
    });

    return dimensionsWithSorts(dimensionsWithoutSorts, def.sortBy);
}
