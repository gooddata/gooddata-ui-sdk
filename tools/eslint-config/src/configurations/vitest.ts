// (C) 2025 GoodData Corporation

import type { IConfiguration } from "../types.js";

const configuration: IConfiguration<"@vitest"> = {
    packages: [
        {
            name: "@vitest/eslint-plugin",
            version: "1.3.5",
        },
    ],
    plugin: "@vitest",
    extends: ["plugin:@vitest/legacy-recommended"],
    rules: {
        "@vitest/expect-expect": "off",
        "@vitest/no-commented-out-tests": "warn",
        "@vitest/valid-title": "warn",
        "@vitest/no-disabled-tests": "warn",
        "@vitest/no-focused-tests": "warn",
        "@vitest/no-identical-title": "warn",
        "@vitest/valid-expect": "warn",
    },
};

export default configuration;
