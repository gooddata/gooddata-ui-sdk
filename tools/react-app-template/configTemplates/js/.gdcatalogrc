{
  "hostname": "https://public-examples.gooddata.com",
  "workspaceId": "demo",
  "output": "src/catalog.ts",
  "backend": "tiger",
  "demo": true
}