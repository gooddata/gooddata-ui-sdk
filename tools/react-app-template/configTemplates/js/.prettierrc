{
    "printWidth": 110,
    "tabWidth": 4,
    "trailingComma": "all"
}
