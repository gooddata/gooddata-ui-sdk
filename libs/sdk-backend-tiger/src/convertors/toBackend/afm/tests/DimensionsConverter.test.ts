// (C) 2020-2022 GoodData Corporation

import {
    defWithDimensions,
    defWithSorting,
    defWithPostProcessing,
    emptyDef,
    IExecutionDefinition,
    MeasureGroupIdentifier,
    newAttributeLocator,
    newAttributeSort,
    newDimension,
    newMeasureSort,
    IPostProcessing,
} from "@gooddata/sdk-model";
import { convertDimensions } from "../DimensionsConverter.js";
import { describe, expect, it } from "vitest";

const SingleDimAttributes = defWithDimensions(emptyDef("test"), newDimension(["localId1", "localId2"]));
const SingleDimAttributesAndMeasure = defWithDimensions(
    emptyDef("test"),
    newDimension(["localId1", "localId2", MeasureGroupIdentifier]),
);
const SingleDimMeasure = defWithDimensions(emptyDef("test"), newDimension([MeasureGroupIdentifier]));
const TwoDimAttrAndMeasures = defWithDimensions(
    emptyDef("test"),
    newDimension(["localId1", "localId2"]),
    newDimension([MeasureGroupIdentifier]),
);
const TwoDimMeasuresAndAttr = defWithDimensions(
    emptyDef("test"),
    newDimension([MeasureGroupIdentifier]),
    newDimension(["localId1", "localId2"]),
);
const TwoDimAttrMeasuresAndAttr = defWithDimensions(
    emptyDef("test"),
    newDimension(["localId1", MeasureGroupIdentifier]),
    newDimension(["localId2"]),
);
const TwoDimAttrAndAttrMeasures = defWithDimensions(
    emptyDef("test"),
    newDimension(["localId2"]),
    newDimension(["localId1", MeasureGroupIdentifier]),
);

describe("convertDimensions", () => {
    const Scenarios: Array<[string, IExecutionDefinition]> = [
        ["single dim definition with just attributes", SingleDimAttributes],
        ["single dim definition with attributes and measure group", SingleDimAttributesAndMeasure],
        ["single dim definition with just measure group", SingleDimMeasure],
        ["two dim definition with attributes in first and measure group in second", TwoDimAttrAndMeasures],
        ["two dim definition with attributes in second and measure group in first", TwoDimMeasuresAndAttr],
        [
            "two dim definition with attributes+measure group in first and attributes in second",
            TwoDimAttrMeasuresAndAttr,
        ],
        [
            "two dim definition with attributes+measure group in second and attributes in first",
            TwoDimAttrAndAttrMeasures,
        ],
    ];

    it.each(Scenarios)("should correctly convert %s", (_desc, def) => {
        expect(convertDimensions(def)).toMatchSnapshot();
    });

    it.each(Scenarios)("should correctly convert %s with attribute sorts", (_desc, def) => {
        const defWithSorts = defWithSorting(def, [newAttributeSort("localId1")]);

        expect(convertDimensions(defWithSorts)).toMatchSnapshot();
    });

    it.each(Scenarios)(
        "should correctly convert %s with measure sorts without attribute locators",
        (_desc, def) => {
            const defWithSorts = defWithSorting(def, [newMeasureSort("testMeasure")]);

            expect(convertDimensions(defWithSorts)).toMatchSnapshot();
        },
    );

    it.each(Scenarios)(
        "should correctly convert %s with measure sorts with attribute locators",
        (_desc, def) => {
            const measureSort = newMeasureSort("testMeasure", "desc", [
                newAttributeLocator("localId1", "PrimaryLabelValue"),
            ]);
            const defWithSorts = defWithSorting(def, [measureSort]);

            expect(convertDimensions(defWithSorts)).toMatchSnapshot();
        },
    );

    it.each(Scenarios)(
        "should correctly convert %s with measure sorts with attribute locators that contain element URIs",
        (_desc, def) => {
            const measureSort = newMeasureSort("testMeasure", "desc", [
                newAttributeLocator("localId1", "/obj/666/elements?id=PrimaryLabelValue"),
            ]);
            const defWithSorts = defWithSorting(def, [measureSort]);

            expect(convertDimensions(defWithSorts)).toMatchSnapshot();
        },
    );

    it.each(Scenarios)("should correctly convert %s with date formats", (_desc, def) => {
        const dateFormats = ["MM/dd/yyyy", "dd/MM/yyyy", "dd-MM-yyyy", "yyyy-MM-dd", "M/d/yy", "dd.MM.yyyy"];
        dateFormats.forEach((dateFormat: string) => {
            const postProcessing: IPostProcessing = { dateFormat };
            const newDef = defWithPostProcessing(def, postProcessing);
            expect(newDef).not.toBe(def);
            expect(newDef.postProcessing?.dateFormat).toBe(dateFormat);
        });
    });
});
