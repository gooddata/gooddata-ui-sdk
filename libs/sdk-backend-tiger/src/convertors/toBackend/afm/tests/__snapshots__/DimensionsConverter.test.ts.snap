// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`convertDimensions > should correctly convert single dim definition with attributes and measure group 1`] = `
[
  {
    "itemIdentifiers": [
      "localId1",
      "localId2",
      "measureGroup",
    ],
    "localIdentifier": "dim_0",
  },
]
`;

exports[`convertDimensions > should correctly convert single dim definition with attributes and measure group with attribute sorts 1`] = `
[
  {
    "itemIdentifiers": [
      "localId1",
      "localId2",
      "measureGroup",
    ],
    "localIdentifier": "dim_0",
    "sorting": [
      {
        "attribute": {
          "attributeIdentifier": "localId1",
          "sortType": "DEFAULT",
        },
      },
    ],
  },
]
`;

exports[`convertDimensions > should correctly convert single dim definition with attributes and measure group with measure sorts with attribute locators 1`] = `
[
  {
    "itemIdentifiers": [
      "localId1",
      "localId2",
      "measureGroup",
    ],
    "localIdentifier": "dim_0",
  },
]
`;

exports[`convertDimensions > should correctly convert single dim definition with attributes and measure group with measure sorts with attribute locators that contain element URIs 1`] = `
[
  {
    "itemIdentifiers": [
      "localId1",
      "localId2",
      "measureGroup",
    ],
    "localIdentifier": "dim_0",
  },
]
`;

exports[`convertDimensions > should correctly convert single dim definition with attributes and measure group with measure sorts without attribute locators 1`] = `
[
  {
    "itemIdentifiers": [
      "localId1",
      "localId2",
      "measureGroup",
    ],
    "localIdentifier": "dim_0",
  },
]
`;

exports[`convertDimensions > should correctly convert single dim definition with just attributes 1`] = `
[
  {
    "itemIdentifiers": [
      "localId1",
      "localId2",
    ],
    "localIdentifier": "dim_0",
  },
]
`;

exports[`convertDimensions > should correctly convert single dim definition with just attributes with attribute sorts 1`] = `
[
  {
    "itemIdentifiers": [
      "localId1",
      "localId2",
    ],
    "localIdentifier": "dim_0",
    "sorting": [
      {
        "attribute": {
          "attributeIdentifier": "localId1",
          "sortType": "DEFAULT",
        },
      },
    ],
  },
]
`;

exports[`convertDimensions > should correctly convert single dim definition with just attributes with measure sorts with attribute locators 1`] = `
[
  {
    "itemIdentifiers": [
      "localId1",
      "localId2",
    ],
    "localIdentifier": "dim_0",
  },
]
`;

exports[`convertDimensions > should correctly convert single dim definition with just attributes with measure sorts with attribute locators that contain element URIs 1`] = `
[
  {
    "itemIdentifiers": [
      "localId1",
      "localId2",
    ],
    "localIdentifier": "dim_0",
  },
]
`;

exports[`convertDimensions > should correctly convert single dim definition with just attributes with measure sorts without attribute locators 1`] = `
[
  {
    "itemIdentifiers": [
      "localId1",
      "localId2",
    ],
    "localIdentifier": "dim_0",
  },
]
`;

exports[`convertDimensions > should correctly convert single dim definition with just measure group 1`] = `
[
  {
    "itemIdentifiers": [
      "measureGroup",
    ],
    "localIdentifier": "dim_0",
  },
]
`;

exports[`convertDimensions > should correctly convert single dim definition with just measure group with attribute sorts 1`] = `
[
  {
    "itemIdentifiers": [
      "measureGroup",
    ],
    "localIdentifier": "dim_0",
  },
]
`;

exports[`convertDimensions > should correctly convert single dim definition with just measure group with measure sorts with attribute locators 1`] = `
[
  {
    "itemIdentifiers": [
      "measureGroup",
    ],
    "localIdentifier": "dim_0",
  },
]
`;

exports[`convertDimensions > should correctly convert single dim definition with just measure group with measure sorts with attribute locators that contain element URIs 1`] = `
[
  {
    "itemIdentifiers": [
      "measureGroup",
    ],
    "localIdentifier": "dim_0",
  },
]
`;

exports[`convertDimensions > should correctly convert single dim definition with just measure group with measure sorts without attribute locators 1`] = `
[
  {
    "itemIdentifiers": [
      "measureGroup",
    ],
    "localIdentifier": "dim_0",
  },
]
`;

exports[`convertDimensions > should correctly convert two dim definition with attributes in first and measure group in second 1`] = `
[
  {
    "itemIdentifiers": [
      "localId1",
      "localId2",
    ],
    "localIdentifier": "dim_0",
  },
  {
    "itemIdentifiers": [
      "measureGroup",
    ],
    "localIdentifier": "dim_1",
  },
]
`;

exports[`convertDimensions > should correctly convert two dim definition with attributes in first and measure group in second with attribute sorts 1`] = `
[
  {
    "itemIdentifiers": [
      "localId1",
      "localId2",
    ],
    "localIdentifier": "dim_0",
    "sorting": [
      {
        "attribute": {
          "attributeIdentifier": "localId1",
          "sortType": "DEFAULT",
        },
      },
    ],
  },
  {
    "itemIdentifiers": [
      "measureGroup",
    ],
    "localIdentifier": "dim_1",
  },
]
`;

exports[`convertDimensions > should correctly convert two dim definition with attributes in first and measure group in second with measure sorts with attribute locators 1`] = `
[
  {
    "itemIdentifiers": [
      "localId1",
      "localId2",
    ],
    "localIdentifier": "dim_0",
    "sorting": [
      {
        "value": {
          "dataColumnLocators": {
            "dim_1": {
              "localId1": "PrimaryLabelValue",
              "measureGroup": "testMeasure",
            },
          },
          "direction": "DESC",
        },
      },
    ],
  },
  {
    "itemIdentifiers": [
      "measureGroup",
    ],
    "localIdentifier": "dim_1",
  },
]
`;

exports[`convertDimensions > should correctly convert two dim definition with attributes in first and measure group in second with measure sorts with attribute locators that contain element URIs 1`] = `
[
  {
    "itemIdentifiers": [
      "localId1",
      "localId2",
    ],
    "localIdentifier": "dim_0",
    "sorting": [
      {
        "value": {
          "dataColumnLocators": {
            "dim_1": {
              "localId1": "PrimaryLabelValue",
              "measureGroup": "testMeasure",
            },
          },
          "direction": "DESC",
        },
      },
    ],
  },
  {
    "itemIdentifiers": [
      "measureGroup",
    ],
    "localIdentifier": "dim_1",
  },
]
`;

exports[`convertDimensions > should correctly convert two dim definition with attributes in first and measure group in second with measure sorts without attribute locators 1`] = `
[
  {
    "itemIdentifiers": [
      "localId1",
      "localId2",
    ],
    "localIdentifier": "dim_0",
    "sorting": [
      {
        "value": {
          "dataColumnLocators": {
            "dim_1": {
              "measureGroup": "testMeasure",
            },
          },
          "direction": "ASC",
        },
      },
    ],
  },
  {
    "itemIdentifiers": [
      "measureGroup",
    ],
    "localIdentifier": "dim_1",
  },
]
`;

exports[`convertDimensions > should correctly convert two dim definition with attributes in second and measure group in first 1`] = `
[
  {
    "itemIdentifiers": [
      "measureGroup",
    ],
    "localIdentifier": "dim_0",
  },
  {
    "itemIdentifiers": [
      "localId1",
      "localId2",
    ],
    "localIdentifier": "dim_1",
  },
]
`;

exports[`convertDimensions > should correctly convert two dim definition with attributes in second and measure group in first with attribute sorts 1`] = `
[
  {
    "itemIdentifiers": [
      "measureGroup",
    ],
    "localIdentifier": "dim_0",
  },
  {
    "itemIdentifiers": [
      "localId1",
      "localId2",
    ],
    "localIdentifier": "dim_1",
    "sorting": [
      {
        "attribute": {
          "attributeIdentifier": "localId1",
          "sortType": "DEFAULT",
        },
      },
    ],
  },
]
`;

exports[`convertDimensions > should correctly convert two dim definition with attributes in second and measure group in first with measure sorts with attribute locators 1`] = `
[
  {
    "itemIdentifiers": [
      "measureGroup",
    ],
    "localIdentifier": "dim_0",
  },
  {
    "itemIdentifiers": [
      "localId1",
      "localId2",
    ],
    "localIdentifier": "dim_1",
    "sorting": [
      {
        "value": {
          "dataColumnLocators": {
            "dim_0": {
              "localId1": "PrimaryLabelValue",
              "measureGroup": "testMeasure",
            },
          },
          "direction": "DESC",
        },
      },
    ],
  },
]
`;

exports[`convertDimensions > should correctly convert two dim definition with attributes in second and measure group in first with measure sorts with attribute locators that contain element URIs 1`] = `
[
  {
    "itemIdentifiers": [
      "measureGroup",
    ],
    "localIdentifier": "dim_0",
  },
  {
    "itemIdentifiers": [
      "localId1",
      "localId2",
    ],
    "localIdentifier": "dim_1",
    "sorting": [
      {
        "value": {
          "dataColumnLocators": {
            "dim_0": {
              "localId1": "PrimaryLabelValue",
              "measureGroup": "testMeasure",
            },
          },
          "direction": "DESC",
        },
      },
    ],
  },
]
`;

exports[`convertDimensions > should correctly convert two dim definition with attributes in second and measure group in first with measure sorts without attribute locators 1`] = `
[
  {
    "itemIdentifiers": [
      "measureGroup",
    ],
    "localIdentifier": "dim_0",
  },
  {
    "itemIdentifiers": [
      "localId1",
      "localId2",
    ],
    "localIdentifier": "dim_1",
    "sorting": [
      {
        "value": {
          "dataColumnLocators": {
            "dim_0": {
              "measureGroup": "testMeasure",
            },
          },
          "direction": "ASC",
        },
      },
    ],
  },
]
`;

exports[`convertDimensions > should correctly convert two dim definition with attributes+measure group in first and attributes in second 1`] = `
[
  {
    "itemIdentifiers": [
      "localId1",
      "measureGroup",
    ],
    "localIdentifier": "dim_0",
  },
  {
    "itemIdentifiers": [
      "localId2",
    ],
    "localIdentifier": "dim_1",
  },
]
`;

exports[`convertDimensions > should correctly convert two dim definition with attributes+measure group in first and attributes in second with attribute sorts 1`] = `
[
  {
    "itemIdentifiers": [
      "localId1",
      "measureGroup",
    ],
    "localIdentifier": "dim_0",
    "sorting": [
      {
        "attribute": {
          "attributeIdentifier": "localId1",
          "sortType": "DEFAULT",
        },
      },
    ],
  },
  {
    "itemIdentifiers": [
      "localId2",
    ],
    "localIdentifier": "dim_1",
  },
]
`;

exports[`convertDimensions > should correctly convert two dim definition with attributes+measure group in first and attributes in second with measure sorts with attribute locators 1`] = `
[
  {
    "itemIdentifiers": [
      "localId1",
      "measureGroup",
    ],
    "localIdentifier": "dim_0",
  },
  {
    "itemIdentifiers": [
      "localId2",
    ],
    "localIdentifier": "dim_1",
    "sorting": [
      {
        "value": {
          "dataColumnLocators": {
            "dim_0": {
              "localId1": "PrimaryLabelValue",
              "measureGroup": "testMeasure",
            },
          },
          "direction": "DESC",
        },
      },
    ],
  },
]
`;

exports[`convertDimensions > should correctly convert two dim definition with attributes+measure group in first and attributes in second with measure sorts with attribute locators that contain element URIs 1`] = `
[
  {
    "itemIdentifiers": [
      "localId1",
      "measureGroup",
    ],
    "localIdentifier": "dim_0",
  },
  {
    "itemIdentifiers": [
      "localId2",
    ],
    "localIdentifier": "dim_1",
    "sorting": [
      {
        "value": {
          "dataColumnLocators": {
            "dim_0": {
              "localId1": "PrimaryLabelValue",
              "measureGroup": "testMeasure",
            },
          },
          "direction": "DESC",
        },
      },
    ],
  },
]
`;

exports[`convertDimensions > should correctly convert two dim definition with attributes+measure group in first and attributes in second with measure sorts without attribute locators 1`] = `
[
  {
    "itemIdentifiers": [
      "localId1",
      "measureGroup",
    ],
    "localIdentifier": "dim_0",
  },
  {
    "itemIdentifiers": [
      "localId2",
    ],
    "localIdentifier": "dim_1",
    "sorting": [
      {
        "value": {
          "dataColumnLocators": {
            "dim_0": {
              "measureGroup": "testMeasure",
            },
          },
          "direction": "ASC",
        },
      },
    ],
  },
]
`;

exports[`convertDimensions > should correctly convert two dim definition with attributes+measure group in second and attributes in first 1`] = `
[
  {
    "itemIdentifiers": [
      "localId2",
    ],
    "localIdentifier": "dim_0",
  },
  {
    "itemIdentifiers": [
      "localId1",
      "measureGroup",
    ],
    "localIdentifier": "dim_1",
  },
]
`;

exports[`convertDimensions > should correctly convert two dim definition with attributes+measure group in second and attributes in first with attribute sorts 1`] = `
[
  {
    "itemIdentifiers": [
      "localId2",
    ],
    "localIdentifier": "dim_0",
  },
  {
    "itemIdentifiers": [
      "localId1",
      "measureGroup",
    ],
    "localIdentifier": "dim_1",
    "sorting": [
      {
        "attribute": {
          "attributeIdentifier": "localId1",
          "sortType": "DEFAULT",
        },
      },
    ],
  },
]
`;

exports[`convertDimensions > should correctly convert two dim definition with attributes+measure group in second and attributes in first with measure sorts with attribute locators 1`] = `
[
  {
    "itemIdentifiers": [
      "localId2",
    ],
    "localIdentifier": "dim_0",
    "sorting": [
      {
        "value": {
          "dataColumnLocators": {
            "dim_1": {
              "localId1": "PrimaryLabelValue",
              "measureGroup": "testMeasure",
            },
          },
          "direction": "DESC",
        },
      },
    ],
  },
  {
    "itemIdentifiers": [
      "localId1",
      "measureGroup",
    ],
    "localIdentifier": "dim_1",
  },
]
`;

exports[`convertDimensions > should correctly convert two dim definition with attributes+measure group in second and attributes in first with measure sorts with attribute locators that contain element URIs 1`] = `
[
  {
    "itemIdentifiers": [
      "localId2",
    ],
    "localIdentifier": "dim_0",
    "sorting": [
      {
        "value": {
          "dataColumnLocators": {
            "dim_1": {
              "localId1": "PrimaryLabelValue",
              "measureGroup": "testMeasure",
            },
          },
          "direction": "DESC",
        },
      },
    ],
  },
  {
    "itemIdentifiers": [
      "localId1",
      "measureGroup",
    ],
    "localIdentifier": "dim_1",
  },
]
`;

exports[`convertDimensions > should correctly convert two dim definition with attributes+measure group in second and attributes in first with measure sorts without attribute locators 1`] = `
[
  {
    "itemIdentifiers": [
      "localId2",
    ],
    "localIdentifier": "dim_0",
    "sorting": [
      {
        "value": {
          "dataColumnLocators": {
            "dim_1": {
              "measureGroup": "testMeasure",
            },
          },
          "direction": "ASC",
        },
      },
    ],
  },
  {
    "itemIdentifiers": [
      "localId1",
      "measureGroup",
    ],
    "localIdentifier": "dim_1",
  },
]
`;
